Firestore/core/test/firebase/firestore/core/event_manager_test.mm

```diff
@@ -0,0 +1,173 @@
+  return QueryListener::Create(std::move(query),
+                               ListenOptions::DefaultOptions(),
+                               NoopViewSnapshotHandler());
+}
+
+class MockEventSource : public core::QueryEventSource {
+ public:
+  MOCK_METHOD1(SetCallback, void(core::QueryEventCallback*));
+  MOCK_METHOD1(Listen, model::TargetId(core::Query));
+  MOCK_METHOD1(StopListening, void(const core::Query&));
+};
+
+TEST(EventManagerTest, HandlesManyListnersPerQuery) {
+  core::Query query = Query("foo/bar");
+  auto listener1 = NoopQueryListener(query);
+  auto listener2 = NoopQueryListener(query);
+
+  MockEventSource mock_event_source;
+  EXPECT_CALL(mock_event_source, SetCallback(_));
+  EventManager event_manager(&mock_event_source);
+
+  EXPECT_CALL(mock_event_source, Listen(query));
+  event_manager.AddQueryListener(listener1);
+
+  EXPECT_CALL(mock_event_source, Listen(_)).Times(0);
+  event_manager.AddQueryListener(listener2);
+
+  EXPECT_CALL(mock_event_source, StopListening(_)).Times(0);
+  event_manager.RemoveQueryListener(listener2);
+
+  EXPECT_CALL(mock_event_source, StopListening(query));
+  event_manager.RemoveQueryListener(listener1);
+}
+
+TEST(EventManagerTest, HandlesUnlistenOnUnknownListenerGracefully) {
+  core::Query query = Query("foo/bar");
+  auto listener = NoopQueryListener(query);
+
+  MockEventSource mock_event_source;
+  EventManager event_manager(&mock_event_source);
+
+  EXPECT_CALL(mock_event_source, StopListening(_)).Times(0);
+  event_manager.RemoveQueryListener(listener);
+}
+
+ViewSnapshot make_empty_view_snapshot(const core::Query& query) {
+  DocumentSet empty_docs{query.Comparator()};
+  // sync_state_changed has to be `true` to prevent an assertion about a
+  // meaningless view snapshot.
+  return ViewSnapshot{query,
+                      empty_docs,
+                      empty_docs,
+                      {},
+                      DocumentKeySet{},
+                      false,
+                      /*sync_state_changed=*/true,
+                      false};
+}
+
+TEST(EventManagerTest, NotifiesListenersInTheRightOrder) {
+  core::Query query1 = Query("foo/bar");
+  core::Query query2 = Query("bar/baz");
+  std::vector<std::string> eventOrder;
+
+  auto listener1 =
+      QueryListener::Create(query1, [&eventOrder](StatusOr<ViewSnapshot>) {
+        eventOrder.push_back("listener1");
+      });
+  auto listener2 =
+      QueryListener::Create(query2, [&eventOrder](StatusOr<ViewSnapshot>) {
+        eventOrder.push_back("listener2");
+      });
+  auto listener3 =
+      QueryListener::Create(query1, [&eventOrder](StatusOr<ViewSnapshot>) {
+        eventOrder.push_back("listener3");
+      });
+
+  MockEventSource mock_event_source;
+  EventManager event_manager(&mock_event_source);
+
+  event_manager.AddQueryListener(listener1);
+  event_manager.AddQueryListener(listener2);
+  event_manager.AddQueryListener(listener3);
+
```

Comment:
I removed these checks intentionally, and the one below, because the behaviour is already tested in `HandlesManyListnersPerQuery`. Adding too many checks distracts reader's attention from the real purpose of this particular case, without adding too much value.

I did what you suggested because I realized `HandlesManyListnersPerQuery` is not testing exactly the same thing, so I'm OK to add them, though I still think checking `event_order` in the last is sufficient for this case.

Reply:
This is reasonable, but keep in mind that any deviation from other SDKs has an associated cost (anyone modifying those tests in the future across platforms would have to stop and think "Why are these tests doing different things? Which one is correct?"). Changing other SDKs to match is additional work as well. Sometimes, this means that for really minor things, leaving them as they are is the lesser evil.

If you feel strongly about this change, please either add a `PORTING NOTE` comment or backport this change to other platforms as well. I'd personally suggest just following the other SDKs in this particular case, because the benefits of the change seem small, but that's not something I feel super strong about.

Reply:
Agree in general. 

The current state is already in sync with other platform actually, Android/Web don't check if `setCallback` is called.